ls1intum/Artemis
# [Enhancement] Fix zero points exercises with and without bonus for all exercise types (manual assessment + complaints)

### Checklist
- [x]  Server: I followed the coding and design guidelines.
- [x]  Client: I followed the coding and design guidelines.

### Motivation and Context


Currently zero point exercises are not completely supported on all exercise types for manual assessment and complaints. This PR introduces the support for this.

### Description

- [x] Extract `getMaxScoreRespectingZeroPointExercises` so that it can be used from other components on server and client
- [x] Implement correct use of maxPoints for assessment of Modeling, Text and FileUpload exercises + complaints
- [x] Implement correct use of maxPoints for assessment of Programming exercises + complaints

### Steps for Testing


Normal exercises:

1. Log in to Artemis
2. Navigate to Course Administration

**Important: test the following 4 cases in 4 different exercises for Programming, Modeling, Text, Quiz and FileUpload:**
 - [x] 10 points + 0 bonus points
 - [x] 10 points + 10 bonus points
 - [x] 0 points + 10 bonus points
 - [x] 0 points + 0 bonus points

Test the following scenario: if a student e.g. has collected 30 points in a course, but the sum of normal points is 20, then the score on the start page of Artemis should show 150%. The score calculation is really critical and has to be absolutely correct!
Also make sure that the instructor page course scores includes the correct calculation based on the sum of normal points and the exported csv file is correct.

3. Assess the exercises 
4. See if everything works as expected
5. Complain and assess the complaint
6. See if everything works as expected

### Test Coverage


Server:

- Exercise.java: 92%
- Result.java: 95%
- AssessmentService.java: 88%
- ProgrammingAssessmentService.java: 92%
- ProgrammingExerciseGradingService.java: 95%

Client:

- scoreChartPreset.ts: 92%
- exercise-utils.ts: 91%
- result-detail.component.ts: 95%
- code-editor-tutor-assessment-container.component: 81%
- text-feedback-conflicts.component.ts: 79%
- modeling-assessment.component.ts: 78.95%
- score-display.component: 76%
- modeling-assessment-editor.component.ts: 72%
- text-assessment-base.component.ts: 71%
- text-submission-assessment.component: 70.11%
- modeling-submission.component.ts:  72.5%
- file-upload-assessment.component.ts: 73.44%

### Screenshots



Tooltip when bonus points are available:
![image](https://user-images.githubusercontent.com/33484318/103211298-18d35080-4908-11eb-9773-d4b66a2ff4a3.png)

Statistics view of an exercise:
![image](https://user-images.githubusercontent.com/33484318/103211730-44a30600-4909-11eb-8248-6a3351e1247e.png)

0 point and 10 bonus point exercise:
![image](https://user-images.githubusercontent.com/33484318/103212353-d52e1600-490a-11eb-93ea-c8589226d3b8.png)

0 point and 0 bonus point exercise (showing only the percentage reached, base is 100 points):
![image](https://user-images.githubusercontent.com/33484318/103212456-18888480-490b-11eb-913f-9935948efb81.png)
